pm: optimize resources

gmarull opened the pull request:
It is well known that PM subsystem has never been optimized in terms of
resource usage. The situation is particularly bad in case the PM runtime
API is enabled. What this patch does is to move the responsibility of PM
resource definition to the device driver like this:

- Device is responsible to define PM resources, using a new set of
  macros: PM_DEVICE_*DEFINE().
- DEVICE_*DEFINE macro accepts a reference to the device PM state, which
  can be obtained using PM_DEVICE_*REF() set of macros. This
  allows device to initialize the dev->pm reference.

This method decouples a bit more PM from devices since devices just keep
a reference to the device PM state. It also means that future PM changes
will have less chances to impact all devices, but only devices that
support PM.

Stats for `west build -b nrf52840dk_nrf52840 samples/hello_world --  -DCONFIG_PM=y -DCONFIG_PM_DEVICE=y -DCONFIG_PM_DEVICE_RUNTIME=y`

main:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       19632 B         1 MB      1.87%
            SRAM:        7360 B       256 KB      2.81%
        IDT_LIST:          0 GB         2 KB      0.00%
```

This PR:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       19188 B         1 MB      1.83%
            SRAM:        6912 B       256 KB      2.64%
        IDT_LIST:          0 GB         2 KB      0.00%
```

gmarull committed:
pm: optimize resource usage

It is well known that PM subsystem has never been optimized in terms of
resource usage. The situation is particularly bad in case the PM runtime
API is enabled. What this patch does is to move the responsability of PM
resource definition to the device like this:

- Device is responsible to define PM resources, using a new set of
  macros: PM_DEVICE_*DEFINE().
- DEVICE_*DEFINE macro accepts a reference to the device PM state, which
  can be obtained using PM_DEVICE_*REF() set of macros. This
  allows device to initialize the dev->pm reference.

This method decouples a bit more PM from devices since devices just keep
a reference to the device PM state. It also means that future PM changes
will have less chances to impact all devices, but only devices that
support PM.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: use new PM macros

Port some drivers to the recently introduced macros to showcase its
usage and be able to do some initial testing (nRF52840).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

ceolin commented:
I think get rid of the pm struct with the when `PM_DISABLE` is passed instead of the callback is a good thing. The other change I have concerns, I think it makes the code more complicated, harder to read and maintain. Instead of jump in optimizations I think we should review the async usage and implementation.

This is not a :-1: but I have doubts if the benefits are greater than the cons.

yperess reviewed (dismissed):
Finally, if we were to add a binding API, we would need a few changes to `subsys/pm` I would add the following to the build:

```
zephyr_sources_ifdef(CONFIG_PM    pm.c)
```

Here I would add an initializer function that will run (probably post kernel). It would iterate through the new section added for the `struct pm_device` and set the `.pm` pointers for all the devices which had PM bindings.

(these are just my thoughts/suggestions, please implement as you see fit)

keith-zephyr reviewed (changes requested):
You also need to update subsys/pm/device.c to change all `if (pm->action_cb == NULL)` checks to `if (pm == NULL)`

gmarull committed:
tests: pm: use new PM macros

Use PM_DEVICE_STATE_DEFINE to define PM state.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

ceolin reviewed (approved):
That is the right direction IMHO. I think the late binding is better but that  can be done on top of this one.

yperess commented:
@gmarull @keith-zephyr I finally got things working, see https://github.com/yperess/zephyr/commit/067f701be674cfd421abd00db3f5bd2b61daa86c for an example. This builds:

```
$ west build -p -b nrf52840dk_nrf52840 samples/subsys/pm/device_pm/
```

gmarull committed:
samples: pm: use new PM macros

Use recently introduced PM macros to define device PM resources.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: interrupt_controller: i/loapic: use DEVICE_DEFINE

Define the device using DEVICE_DEFINE macro, so that a single option can
be used regardless of PM being enabled or not.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
device: remove PM capability from SYS_DEVICE_DEFINE

The macro already mentions in the docstrings that PM is not supported:

"Invokes DEVICE_DEFINE() with no power management support".

This patch removed the PM entry from the macro and ajusts its uses.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

keith-zephyr commented on include/device.h:
```diff
@@ -667,9 +652,7 @@ static inline bool device_is_ready(const struct device *dev)
  */
 #define Z_DEVICE_STATE_DEFINE(node_id, dev_name, pm_action_cb)		\
```
pm_action_cb is no longer used by `Z_DEVICE_STATE_DEFINE()`.  pm_action_cb should also be removed from `Z_DEVICE_DEFINE_PRE()`

gmarull replied:
fixed

keith-zephyr commented on drivers/display/display_st7735r.c:
```diff
@@ -573,7 +573,10 @@ static const struct display_driver_api st7735r_api = {
+										\
+	PM_DEVICE_DT_INST_DEFINE(inst, st7735r_pm_action);			\
+										\
+	DEVICE_DT_INST_DEFINE(inst, st7735r_init, PM_DEVICE_DT_INST_REF(int),	\
```
Should be `PM_DEVICE_DT_INST_REF(inst)` ("int" -> "inst")

gmarull replied:
fixed

keith-zephyr commented on drivers/sensor/ina219/ina219.c:
```diff
@@ -302,9 +302,11 @@ static const struct sensor_driver_api ina219_api = {
 	DEVICE_DT_INST_DEFINE(n,				\
 			      ina219_init,			\
-			      ina219_pm_action,			\
+			      PM_DEVICE_DT_INST_REF(n),	\
```
Align the continuation character.

gmarull replied:
fixed

keith-zephyr commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -488,9 +488,11 @@ static int lis2mdl_pm_action(const struct device *dev,
 	DEVICE_DT_INST_DEFINE(inst,					\
 			    lis2mdl_init,				\
-			    lis2mdl_pm_action,				\
+			    PM_DEVICE_DT_INST_REF(n),			\
```
Should be `PM_DEVICE_DT_INST_REF(inst)`

gmarull replied:
fixed

keith-zephyr commented on drivers/sensor/lm77/lm77.c:
```diff
@@ -417,8 +411,10 @@ static int lm77_pm_action(const struct device *dev,
+									\
 	DEVICE_DT_INST_DEFINE(n, lm77_init,				\
-			      LM77_PM_ACTION_CB,			\
+			      PM_DEVICE_DT_INST_REF(n),		\
```
Alignment.

gmarull commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -1025,9 +1025,11 @@ static int fdc2x1x_init(const struct device *dev)
 	DEVICE_DT_INST_DEFINE(n,					   \
 			      fdc2x1x_init,				   \
-			      fdc2x1x_device_pm_action,			   \
+			      PM_DEVICE_DT_INST_REF(n),		   \
```
fixed

gmarull commented on drivers/interrupt_controller/intc_ioapic.c:
```diff
@@ -495,10 +495,5 @@ static void IoApicRedUpdateLo(unsigned int irq,
-#else
-SYS_INIT(ioapic_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
-#endif
+DEVICE_DEFINE(ioapic, "ioapic", ioapic_init, ioapic_pm_action, NULL, NULL,
```
it will compile since the `pm_action_cb` will basically be ignored by `DEVICE_DEFINE` (to be precise, `Z_DEVICE_STATE_PM_INIT` is no-op).

keith-zephyr commented on include/device.h:
```diff
@@ -731,23 +713,15 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
+#define Z_DEVICE_DEFINE(node_id, dev_name, drv_name, init_fn, pm_device,\
 			data_ptr, cfg_ptr, level, prio, api_ptr, ...)	\
-	Z_DEVICE_DEFINE_PRE(node_id, dev_name, pm_action_cb, __VA_ARGS__) \
+	Z_DEVICE_DEFINE_PRE(node_id, dev_name, __VA_ARGS__) \
```
Align `\`

keith-zephyr commented on include/pm/device.h:
```diff
@@ -144,8 +144,115 @@ struct pm_device {
+ *
+ * @see #PM_DEVICE_DT_DEFINE, #PM_DEVICE_DEFINE
+ */
+#define PM_DEVICE_DT_INST_DEFINE(idx, action_cb)			\
```
Nit - 2nd parameter should be `pm_action_cb`

keith-zephyr commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -630,9 +630,10 @@ static int gpio_stm32_init(const struct device *dev)
 		.pclken = { .bus = __bus, .enr = __cenr }		       \
 	};								       \
 	static struct gpio_stm32_data gpio_stm32_data_## __suffix;	       \
+	PM_DEVICE_DT_INST_DEFINE(__node, gpio_stm32_pm_action);		       \
```
Should be `PM_DEVICE_DT_DEFINE()`

keith-zephyr commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -630,9 +630,10 @@ static int gpio_stm32_init(const struct device *dev)
 	DEVICE_DT_DEFINE(__node,					       \
 			    gpio_stm32_init,				       \
-			    gpio_stm32_pm_action,			       \
+			    PM_DEVICE_DT_INST_REF(__node),		       \
```
Should be `PM_DEVICE_DT_REF()`

gmarull replied:
fixed

gmarull commented:
@yperess can you re-review?

gmarull committed:
tests: posix: common: increase main stack for TLS tests

Increase main stack size (in 128 bytes) to make these tests run
successfully on qemu_x86 platform.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
Confirmed CI passes, dropped buildkite resorce increase patch.
